posts/2021/09/hypothesis-array-api.ipynb

```diff
@@ -0,0 +1,940 @@
+    "    result = xp.empty(x.size, dtype=x.dtype)\n",
+    "    result[0] = x[0]\n",
+    "    for i in range(1, x.size):\n",
+    "        result[i] = result[i - 1] + x[i]\n",
+    "        \n",
+    "    return result"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ed669b65",
+   "metadata": {},
+   "source": [
+    "The all-important\n",
+    "[`__array_namespace__()`](https://data-apis.org/array-api/latest/API_specification/array_object.html#method-array-namespace) method\n",
+    "allows array-consuming methods to get the array's respective Array API module.\n",
+    "Conventionally we assign it to the variable `xp`.\n",
+    "\n",
+    "From there you just need to rely on the guarantees of the Array API\n",
+    "to support NumPy, TensorFlow, PyTorch, CuPy, etc.\n",
+    "all in one simple method!"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6e824f0f",
+   "metadata": {},
+   "source": [
+    "## Good ol' unit tests\n",
+    "\n",
+    "I hope you'd want write some tests at some point 😉\n",
+    "\n",
+    "We can import NumPy's Array API implementation\n",
+    "and test with that for now,\n",
+    "although in the future it'd be a good idea to try other implementations\n",
+    "(see [related Hypothesis issue](https://github.com/HypothesisWorks/hypothesis/issues/3085))."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "cb1e47fe-863a-4ff6-913a-572da5985d04",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\u001b[32m.\u001b[0m\u001b[32m                                                                        [100%]\u001b[0m\n",
+      "\u001b[32m\u001b[32m\u001b[1m1 passed\u001b[0m\u001b[32m in 0.01s\u001b[0m\u001b[0m\n"
+     ]
+    }
+   ],
+   "source": [
+    "from numpy import array_api as nxp\n",
+    "\n",
+    "def test_cumulative_sums():\n",
+    "    x = nxp.asarray([0, 1, 2, 3, 4])\n",
+    "    assert nxp.all(cumulative_sums(x) == nxp.asarray([0, 1, 3, 6, 10]))\n",
+    "    \n",
+    "ipytest.run()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "098c6b37",
+   "metadata": {},
+   "source": [
+    "I would probably write a [parametrized](https://docs.pytest.org/en/stable/parametrize.html) test here\n",
+    "and write cases to cover all the interesting scenarios I can think of.\n",
+    "Whatever we do,\n",
+    "we will definitely miss some edge cases.\n",
+    "What if we could catch bugs\n",
+    "we would never think of ourselves?"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "fd6530d8-3072-464e-8322-a51a877f5b70",
+   "metadata": {},
+   "source": [
+    "## Testing our assumptions with Hypothesis\n",
+    "\n",
+    "Hypothesis is a property-based testing library.\n",
+    "To copy from their excellent [docs](https://hypothesis.readthedocs.io/en/latest/index.html)...\n",
+    "\n",
+    "<!-- I would put this in quote block, but it looks bad with the blog's style -->\n",
```

Comment:
Hm maybe we can put it in italics then?

Reply:
I tried this and after getting it working (had to go the full HTML tags route), I didn't really like it heh. I tried playing around with the CSS and HTML tags, but couldn't get anything nice out of it. My current change seems like the best "easy" solution, before I spend any more time on this haha.